### PR TITLE
[Convdiff] Adding new mount method in move particle utility with a provided cell size

### PR DESCRIPTION
**Pull request by ipouplana**
**Description**
A new mount method has been added in the move_particle_utility.h to allow providing a cell size for the bins object. The previous mount method is unaltered. 

**Changelog**
- Added a new mount method that requires the cell size of the bins object to be called.
- Correctly initializing non-historical elemental variables to avoid memory errors.

### CODE REVIEW

File: applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: Why changing to the `SetValue`? Isn't it initialized before?

File: applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]:
```suggestion
		void MountBin(
		    const bool rCellSizeProvided=false,
		    const double rCellSize=1.0)
```
It is customary to pass simple types as copy. Also minor style guide suggestion.
Besides this, I think that we could only use the `rCellSize` parameter defaulted to zero to then check `if (rCellSize > 0.0)`.

File: applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h
Comment (rubenzorrilla) [inline applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: Same set/get comment.

**Comment by RiccardoRossi** [comment]
guys, the default should not be 1.0, otherwise it would be dimensionally dependent.

instead you should have an option which makes it to work with automatic definiton (as now), OR with a provided size...that is, i would provide a version with one single paramter, and one with 2

**Review by RiccardoRossi — DISMISSED** [review]
you should not default to 1.0

**Comment by rubenzorrilla** [comment]
I was just making implementation related comments. As I've no idea about the physics behind this utility I let @RiccardoRossi to follow this.

File: applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h
Comment (ipouplana) [reply applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: No. And when I ran it in FullDebug it crashed...

Comment (ipouplana) [reply applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h:177]: Well, I see now, it is initialized before... I don't know why it crashed in FullDebug then...

Comment (ipouplana) [reply applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: I see it... The variables are initialized as **nodal** non-historical variables, but they are being used by **elements**...

Comment (ipouplana) [reply applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: I just corrected the initialization

File: applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h
Comment (ipouplana) [reply applications/ConvectionDiffusionApplication/custom_utilities/move_particle_utility.h]: I created a new method with a cell size as argument, instead of modifying the existing one

**Comment by ipouplana** [comment]
Hi @RiccardoRossi , I just created a new method with a cell size as argument, instead of modifying the existing one. This way there is no need to write any default cell size.

**Comment by ipouplana** [comment]
Hi @RiccardoRossi, could you take a look at this ?

File: applications/ConvectionDiffusionApplication/python_scripts/move_particle_utility_process.py
Comment (miguelmaso) [inline applications/ConvectionDiffusionApplication/python_scripts/move_particle_utility_process.py:30]: good catch!

**Comment by ipouplana** [comment]
Ping @RiccardoRossi

**Comment by RiccardoRossi** [comment]
Hi @ipouplana you can even preseve the same name. I mean

* MountBin()
* MountBin(CellSize)

would do.
the problem is that you should not assign a default value.

**Comment by ipouplana** [comment]
> Hi @ipouplana you can even preseve the same name. I mean
> 
> * MountBin()
> * MountBin(CellSize)
> 
> would do.
> the problem is that you should not assign a default value. I mean they need to be two methods with two different signatures

Ok, but then wouldn't there be problems when exporting these two methods with the same name to python ?

**Comment by RiccardoRossi** [comment]
you need to have a helper method in the exporting, but it is not a big deal. Take a look for example at the Node interface or at the model_part

**Comment by ipouplana** [comment]
> you need to have a helper method in the exporting, but it is not a big deal. Take a look for example at the Node interface or at the model_part

I've looked at the [Node](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_node_to_python.cpp ) interface and at the [ModelPart](https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp) interface and I still don't know exactly what I am supposed to do ?
Can you explain me exactly how to add a helper method in the exporting ? (sorry but I've never done this before)

**Comment by RiccardoRossi** [comment]
look here

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp#L90

in this function you have 6 overloads (L90, L103, L117 ...) 

you add them in line 

https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp#L1013

does it serve as explanation?

**Comment by ipouplana** [comment]
> look here
> 
> https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp#L90
> 
> in this function you have 6 overloads (L90, L103, L117 ...)
> 
> you add them in line
> 
> https://github.com/KratosMultiphysics/Kratos/blob/master/kratos/python/add_model_part_to_python.cpp#L1013
> 
> does it serve as explanation?

Yes, thank you!

**Comment by miguelmaso** [comment]
@ipouplana done. Not sure if is it ok with the template. If there is a problem, revert it

**Comment by ipouplana** [comment]
> @ipouplana done. Not sure if is it ok with the template. If there is a problem, revert it

I was about to push the same changes... Except for using unsigned int instead of int in the template. I guess it should work like this.

**Comment by ipouplana** [comment]
Thank you @miguelmaso !

**Comment by rubenzorrilla** [comment]
> you need to have a helper method in the exporting, but it is not a big deal. Take a look for example at the Node interface or at the model_part

Just mentioning that a lambda would be cleaner than a helper method.

**Comment by ipouplana** [comment]
> > you need to have a helper method in the exporting, but it is not a big deal. Take a look for example at the Node interface or at the model_part
> 
> Just mentioning that a lambda would be cleaner than a helper method.

Could you point me out an example of this ?

**Comment by miguelmaso** [comment]
> > you need to have a helper method in the exporting, but it is not a big deal. Take a look for example at the Node interface or at the model_part
> 
> Just mentioning that a lambda would be cleaner than a helper method.

pybind::overload_cast will be even cleaner (but it needs c++14 #6611)

**Comment by ipouplana** [comment]
Ping @RiccardoRossi

**Comment by ipouplana** [comment]
@RiccardoRossi or @rubenzorrilla could you take a look at this and decide whether to approve or to ask for more changes ?

**Comment by rubenzorrilla** [comment]
As I mentioned, I've no idea about this implementation so I'd prefer @RiccardoRossi to take the decision on this. About the Python export, it was just a minor comment, current solution is technically correct so I've no problems with it.

**Review by miguelmaso — APPROVED** [review]
Approving since it is a very simple PR and the requested changes have been applied.
Please @ipouplana, wait a few days before merge in case @RiccardoRossi or @rubenzorrilla request more changes or have some concern.

**Comment by rubenzorrilla** [comment]
Fine for me. If @miguelmaso agrees with the changes I think you can go ahead without @RiccardoRossi 's approval.

**Pull request merged**
